src/teleport.cpp

```diff
@@ -0,0 +1,92 @@
+#include "teleport.h"
+
+#include "character.h"
+#include "avatar.h"
+#include "creature.h"
+#include "player.h"
+#include "game.h"
+#include "map.h"
+#include "messages.h"
+#include "point.h"
+
+
+const efftype_id effect_teleglow( "teleglow" );
+
+bool teleport::teleport( Creature &critter, int min_distance, int max_distance, bool safe,
+                         bool add_teleglow )
+{
+    if( min_distance > max_distance ) {
+        debugmsg( "ERROR: Function teleport::teleport called with invalid arguments." );
+        return false;
+    }
+
+    const bool c_is_u = &critter == &g->u;
+    player *const p = critter.as_player();
+    int tries = 0;
+    tripoint origin = critter.pos();
+    tripoint new_pos = tripoint_zero;
+    do {
+        int rangle = rng( 0, 360 );
+        int rdistance = rng( min_distance, max_distance );
+        new_pos.x = origin.x + rdistance * cos( rangle );
+        new_pos.y = origin.y + rdistance * sin( rangle );
+        tries++;
+    } while( g->m.impassable( new_pos ) && tries < 20 );
+    //handles teleporting into solids.
+    if( g->m.impassable( new_pos ) ) {
+        if( safe ) {
+            if( c_is_u ) {
+                add_msg( m_bad, _( "You cannot teleport safely" ) );
+            }
+            return false;
+        }
+        critter.apply_damage( nullptr, bp_torso, 9999 );
+        if( c_is_u ) {
+            g->events().send<event_type::teleports_into_wall>( p->getID(), g->m.obstacle_name( new_pos ) );
+            add_msg( m_bad, _( "You die after teleporting into a solid." ) );
+        }
+        critter.check_dead_state();
+
+    }
+    //handles telefragging other creatures
+    if( Creature *const poor_soul = g->critter_at<Creature>( new_pos ) ) {
+        if( safe ) {
+            if( c_is_u ) {
+                add_msg( m_bad, _( "You cannot teleport safely." ) );
+            }
+            return false;
+        } else {
+            const bool poor_soul_is_u = ( poor_soul == &g->u );
```

Comment:
```suggestion
            const bool poor_soul_is_u = &poor_soul == &g->u;
```

Reply:
Current code  works, I'll keep it unchanged since testing until a monster telefrags you is a real chore.